File: src/Maximum subarray sum 2/Code/solution.py
from collections import deque

class Solution:
    def maxSubarrSum(self, arr, a, b):
        n = len(arr)
        if n == 0:
            return 0

        # prefix sums: prefix[i] = sum of first i elements (arr[0..i-1])
        prefix = [0] * (n + 1)
        for i in range(1, n + 1):
            prefix[i] = prefix[i-1] + arr[i-1]

        dq = deque()    # will store indices with increasing prefix values
        ans = -10**30

        for r in range(1, n + 1):
            leftBound = max(0, r - b)
            newIdx = r - a

            # remove indices out of the window
            while dq and dq[0] < leftBound:
                dq.popleft()

            # insert new index while maintaining increasing prefix values
            if newIdx >= 0:
                while dq and prefix[dq[-1]] >= prefix[newIdx]:
                    dq.pop()
                dq.append(newIdx)

            if dq:
                ans = max(ans, prefix[r] - prefix[dq[0]])

        return ans

File: src/Maximum subarray sum 2/Code/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    // Use long long to avoid overflow of sums
    long long maxSubarrSum(const vector<int>& arr, int a, int b) {
        int n = (int)arr.size();
        if (n == 0) return 0;
        // prefix[i] = sum of first i elements (arr[0..i-1])
        vector<long long> prefix(n + 1, 0);
        for (int i = 1; i <= n; ++i) prefix[i] = prefix[i-1] + (long long)arr[i-1];

        deque<int> dq;               // will store indices of prefix in increasing prefix value order
        long long ans = LLONG_MIN;

        for (int r = 1; r <= n; ++r) {
            int leftBound = max(0, r - b);  // smallest allowed left index
            int newIdx   = r - a;           // index that becomes available at this r (if >= 0)

            // remove indices that are out of the window [leftBound, r-a]
            while (!dq.empty() && dq.front() < leftBound) dq.pop_front();

            // if a left index becomes available, insert it maintaining increasing prefix values
            if (newIdx >= 0) {
                while (!dq.empty() && prefix[dq.back()] >= prefix[newIdx]) dq.pop_back();
                dq.push_back(newIdx);
            }

            // current best for subarray ending at r-1 uses min prefix in deque
            if (!dq.empty()) {
                ans = max(ans, prefix[r] - prefix[dq.front()]);
            }
        }
        return ans;
    }
};
